zerver/views/message_fetch.py

```diff
@@ -263,22 +263,28 @@ def messages_in_narrow_backend(
         )
     )
 
-    builder = NarrowBuilder(user_profile, column("message_id", Integer), user_profile.realm)
-    if narrow is not None:
-        for term in narrow:
-            query = builder.add_term(query, term)
+    inner_msg_id_col = column("message_id", Integer)
+    query, is_search = add_narrow_conditions(
+        user_profile=user_profile,
+        inner_msg_id_col=inner_msg_id_col,
+        query=query,
+        narrow=narrow,
+        is_web_public_query=False,
+        realm=user_profile.realm,
+    )
+
+    if not is_search:
+        # `add_narrow_conditions` adds the following columns only if narrow has search operands.
+        query = query.add_columns(topic_column_sa(), column("rendered_content", Text))
 
     search_fields = {}
     with get_sqlalchemy_connection() as sa_conn:
-        for row in sa_conn.execute(query).fetchall():
-            message_id = row._mapping["message_id"]
-            topic_name = row._mapping[DB_TOPIC_NAME]
-            rendered_content = row._mapping["rendered_content"]
-            if "content_matches" in row._mapping:
-                content_matches = row._mapping["content_matches"]
-                topic_matches = row._mapping["topic_matches"]
-            else:
-                content_matches = topic_matches = []
+        for row in sa_conn.execute(query).mappings():
+            message_id = row["message_id"]
+            topic_name: str = row[DB_TOPIC_NAME]
+            rendered_content: str = row["rendered_content"]
+            content_matches = row.get("content_matches", [])
+            topic_matches = row.get("topic_matches", [])
```

Comment:
Thanks for the update. Just a nit: should we inline these to the `get_search_fields`, so that we don't need to define these variables?